nixos/modules/services/web-apps/fusion.nix

```diff
@@ -0,0 +1,144 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.fusion;
+in
+{
+  options.services.fusion = {
+    enable = lib.mkEnableOption "Fusion, a lightweight, self-hosted friendly RSS aggregator and reader";
+
+    host = lib.mkOption {
+      type = lib.types.str;
+      default = "0.0.0.0";
+      example = "[::1]";
+      description = "The address that Fusion should listen on.";
+    };
+
+    port = lib.mkOption {
+      type = lib.types.port;
+      default = 8080;
+      description = "The port that Fusion should listen on.";
+    };
+
+    passwordFile = lib.mkOption {
+      type = lib.types.path;
+      example = "/run/keys/fusion-password";
+      description = ''
+        A file containing the password for the Fusion web interface.
+
+        This should be a path pointing to a file within a secure directory and NEVER
+        in the Nix store (which is world-readable)!
+      '';
+    };
+
+    tls = lib.mkOption {
+      type = lib.types.nullOr (
+        lib.types.submodule {
+          options = {
+            cert = lib.mkOption {
+              type = lib.types.path;
+              description = "Path to TLS certificate";
+            };
+            key = lib.mkOption {
+              type = lib.types.path;
+              description = "Path to TLS key";
+            };
+          };
+        }
+      );
+      default = null;
+      description = ''
+        The paths to the TLS certificate and key files for Fusion.
+
+        If these options are set, then Fusion can only be accessed through a secure
+        TLS connection. If you are using a reverse proxy like Nginx to handle HTTPS,
+        please leave these unset.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.fusion = {
+      description = "Fusion, a lightweight, self-hosted friendly RSS aggregator and reader";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+
+      environment = {
+        HOST = cfg.host;
+        PORT = toString cfg.port;
+        DB = "/var/lib/fusion/sqlite.db";
+
+        TLS_CERT = lib.mkIf (cfg.tls != null) cfg.tls.cert;
+        TLS_KEY = lib.mkIf (cfg.tls != null) cfg.tls.key;
+      };
+
+      script = ''
+        export PASSWORD=$(cat $CREDENTIALS_DIRECTORY/fusion)
```

Comment:
Isn't this exposing the password in ps?

Reply:
I don't think so? This is run in a run script that's only run by the systemd user, and $CREDENTIALS_DIRECTORY isn't visible anywhere else (see [systemd docs](https://systemd.io/CREDENTIALS/).

It's not optimal, but you can argue that a program that's written to only source its password from an unencrypted environmental variable is already unsafe anyway 🤷🏼‍♀️

Reply:
This is used as an example in the [upstream docs](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#LoadCredential=ID:PATH):

> In order to reference the path a credential may be read from within a ExecStart= command line use "${CREDENTIALS_DIRECTORY}/mycred", e.g. "ExecStart=cat ${CREDENTIALS_DIRECTORY}/mycred"

But next they also give an example of

> In order to reference the path a credential may be read from within a Environment= line use "%d/mycred", e.g. "Environment=MYCREDPATH=%d/mycred".

That may be better to apply here rather than doing it in the script -- assuming it works

Reply:
The thing is that fusion doesn't expect a *path* to the credential, but the whole plain password itself. This is why I mentioned that honestly wanting any amount of actual security on this thing is somewhat futile since it only works with plain passwords 100% visible and readable as an environment variable

Reply:
> This is why I mentioned that honestly wanting any amount of actual security on this thing is somewhat futile since it only works with plain passwords 100% visible and readable as an environment variable

I wouldn't say it's *that* bad, as this environment variable is at least scoped to the service script (compared to `ps` info being available globally on many machines)

I don't *think* this would actually make the secrets contents appear in `ps` though, only the directory -- which doesn't matter as it should have restricted perms. If we really want to avoid cat though, we could use this instead

```bash
PASSWORD="$(< "$CREDENTIALS_DIRECTORY"/fusion)"
export PASSWORD
```